test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,201 @@
+import 'package:checks/checks.dart';
+import 'package:test/scaffolding.dart';
+import 'package:zulip/api/model/events.dart';
+import 'package:zulip/api/model/model.dart';
+import 'package:zulip/model/recent_senders.dart';
+import '../example_data.dart' as eg;
+import 'recent_senders_checks.dart';
+
+void checkMatchesMessages(List<Message> messages, RecentSenders model) {
+  final Map<int, Map<int, MessageIdTracker>> expectedStreamSenders = {};
+  final Map<int, Map<String, Map<int, MessageIdTracker>>> expectedTopicSenders = {};
+  for (final message in messages) {
+    if (message is! StreamMessage) continue;
+
+    final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
+
+    // Track message in stream.
+    var sendersMap = expectedStreamSenders[streamId] ??= {};
+    var idTracker = sendersMap[senderId] ??= MessageIdTracker();
+    idTracker.add(messageId);
+
+    // Track message in topic.
+    final topicsMap = expectedTopicSenders[streamId] ??= {};
+    sendersMap = topicsMap[topic] ??= {};
+    idTracker = sendersMap[senderId] ??= MessageIdTracker();
+    idTracker.add(messageId);
+  }
+  check(model)
+    ..streamSenders.deepEquals(expectedStreamSenders)
+    ..topicSenders.deepEquals(expectedTopicSenders);
+}
+
+void main() {
+  group('RecentSenders', () {
+    test('starts with empty stream and topic senders', () {
+      final model = RecentSenders();
+      checkMatchesMessages([], model);
+    });
+
+    test('just process a stream message', () {
+      final model = RecentSenders();
+      final dmMessage = eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]);
+      final streamMessage = eg.streamMessage(id: 100, sender: eg.user(userId: 10),
+        stream: eg.stream(streamId: 1), topic: 'topic');
+
+      model.handleMessage(dmMessage);
+      model.handleMessage(streamMessage);
+      checkMatchesMessages([streamMessage], model);
+    });
+
+    test('add a batch of messages', () {
```

Comment:
This should test adding a batch when the data structure isn't empty.

In particular let's exercise all the different cases that appear in the logic of the code being tested: the new messages for a given ID tracker are all at the start, or they're all at the end, or they're interspersed among the existing messages. Also cover where some of the new messages are already present.